dowhy/causal_identifiers/efficient_backdoor.py

```diff
@@ -0,0 +1,346 @@
+import networkx as nx
+import numpy as np
+import warnings
+
+
+EXCEPTION_NO_ADJ = "An adjustment set formed by observable variables does not exist"
+EXCEPTION_COND_NO_OPT = "Conditions to guarantee the existence of an optimal adjustment set are not satisfied"
+
+
+class EfficientBackdoor:
+    """
+    Implements methods for finding optimal backdoor sets.
+    """
+
+    def __init__(self, graph, conditional_node_names=None, costs=None):
+        self.graph = graph
+        if costs is None:
+            # If no costs are passed, use uniform costs
+            costs = [(node, {"cost": 1}) for node in self.graph._graph.nodes]
+        assert all([tup['cost'] > 0 for _, tup in costs]), "All costs must be positive"
+        self.graph._graph.add_nodes_from(costs)
+        self.observed_nodes = set(
+            [
+                node
+                for node in self.graph._graph.nodes
+                if self.graph._graph.nodes[node]["observed"] == "yes"
+            ]
+        )
+        if conditional_node_names is None:
+            conditional_node_names = []
+        assert set(conditional_node_names).issubset(self.observed_nodes), "Some conditional variables are not marked as observed"
+        self.conditional_node_names = conditional_node_names
+
+    def ancestors_all(self, nodes):
+        """Returns a set with all ancestors of nodes
+
+        Parameters
+        ----------
+        nodes : list
+           A list of nodes in the graph
+
+        Returns
+        ----------
+        ancestors: set
+
+        Notes
+        -----
+        A node is always an ancestor of itself.
+        """
+        ancestors = set()
+
+        for node in nodes:
+            ancestors_node = nx.ancestors(self.graph._graph, node)
+            ancestors = ancestors.union(ancestors_node)
+
+        ancestors = ancestors.union(set(nodes))
+
+        return ancestors
+
+    def backdoor_graph(self, G):
```

Comment:
is this similar to `self._graph.do_surgery`? did not understand the functionality

Reply:
It's very similar. The only difference is that this only removes edges that lie on some causa path between treatment and outcome, rather than all edges pointing out of the treatment. This is the "proper backdoor graph" introduced [here](https://arxiv.org/pdf/1803.00116.pdf).